{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting regex\n",
      "  Downloading regex-2021.4.4-cp38-cp38-manylinux2014_x86_64.whl (733 kB)\n",
      "\u001b[K     |████████████████████████████████| 733 kB 15.3 MB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: regex\n",
      "Successfully installed regex-2021.4.4\n"
     ]
    }
   ],
   "source": [
    "!pip install regex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import regex as re\n",
    "import operator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Treshold coverage for calling\n",
    "cov_threshold = 2\n",
    "\n",
    "# Variant allele frequency coverage for calling\n",
    "vaf_threshold = 0.2\n",
    "\n",
    "# Threshlod to call the variant heterozygous (GT = 0/1) which means that only one chromosome has found variant\n",
    "# We set here same threshold as for variant allele frequency\n",
    "het_threshold = vaf_threshold\n",
    "\n",
    "# Threshlod to call the variant homozygous (GT = 1/1) which means that both chromosomes have found variant\n",
    "hom_threshold = 0.8\n",
    "\n",
    "# Threshold for quality of bases. If mean base quality is below this threshold we set filter tag for this variant\n",
    "qual_threshold = 40\n",
    "\n",
    "# These two types of mutations (deletions and insertions) are commented\n",
    "# but we gave an regex example how these should be parsed\n",
    "\n",
    "#deletion_reg = re.compile(\"-[0-9]+[ACGTNacgtn]+\")\n",
    "#insertion_reg = re.compile(\"\\+[0-9]+[ACGTNacgtn]+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alternative_base(reads):\n",
    "    \n",
    "    \"\"\"Function for counting alternative bases from reads part from pileup file.\n",
    "    More about pileup which is created from BAM file can be found here: \n",
    "    https://en.wikipedia.org/wiki/Pileup_format\n",
    "    Pileup is created out of BAM file. \n",
    "    Example task where pileup file is created can be found here: \n",
    "    https://cgc.sbgenomics.com/u/ana_djukic/matf-2021/tasks/d8563fba-1576-4755-bdb5-e26251dcdab0/\"\"\"\n",
    "    \n",
    "    alt_dict = {} # create dict structure where keys will be variants and values number of each variant found on specific position\n",
    "    alt_base = \"\" # variable for storing info about variant found in pileup\n",
    "    alt_alleles = 0 # variable for storing info how many variants are found in pileup\n",
    "    \n",
    "    # For example pileup reads are: chr1 T 22 ..^c..,,.,.,.C.,,,.,..c. \n",
    "    # This means that we have two variants, since characater after sign \"^\" gives mapping quality not the alternative bases\n",
    "    alt_base_reg = re.compile(\"(?<!\\^)[ACTGNactgn]\") \n",
    "    alt_bases_list = [x.upper() for x in alt_base_reg.findall(reads)]  # upper all string bases in pileup \n",
    "    \n",
    "    # Check if there is mutation and if \"N\" isn't only one in list\n",
    "    if alt_bases_list != [] and \"N\" not in set(alt_bases_list):\n",
    "        if \"A\" in alt_bases_list:\n",
    "            # Count number of mutations \"A\"\n",
    "            alt_dict[\"A\"] = alt_bases_list.count(\"A\")\n",
    "        elif \"C\" in alt_bases_list:\n",
    "            # Count number of mutations \"C\"\n",
    "            alt_dict[\"C\"] = alt_bases_list.count(\"C\")\n",
    "        elif \"G\" in alt_bases_list:\n",
    "            # Count number of mutations \"G\"\n",
    "            alt_dict[\"G\"] = alt_bases_list.count(\"G\")\n",
    "        elif \"T\" in alt_bases_list:\n",
    "            # Count number of mutations \"T\"\n",
    "            alt_dict[\"T\"] = alt_bases_list.count(\"T\")\n",
    "        \n",
    "        # Take the maximum number of alternative bases as real alternative base.\n",
    "        # Note: Here, we only took maximum, but different algorithms use different statistics to calculate this\n",
    "        alt_base = max(alt_dict, key = alt_dict.get)\n",
    "        # When calculated just take alt base from the created dict\n",
    "        alt_alleles = alt_dict[alt_base]\n",
    "        \n",
    "    return alt_base, alt_alleles\n",
    "\n",
    "def calculate_vaf(num_alt_bases, coverage):\n",
    "    \"\"\"Calculation of Variant allele frequency from\n",
    "    number of alternative bases and coverage.\"\"\"\n",
    "    # Example, pileup reads are: chr1 T 22 ..^c..,,.,.,.C.,,,.,..c. \n",
    "    # Here we have 22 reads aligned on that position, and 2 alternative bases, which gives us VAF: 2/22\n",
    "    vaf = round((num_alt_bases/coverage),2)\n",
    "    \n",
    "    return vaf\n",
    "\n",
    "def genotype(vaf, het_threshold, hom_threshold):\n",
    "    \"\"\"Genotype variants based on thresholds we set as global variables\"\"\"\n",
    "    if (vaf > hom_threshold):\n",
    "        GT = \"1/1\"\n",
    "    elif (vaf < hom_threshold and vaf > het_threshold):\n",
    "        GT = \"0/1\"\n",
    "    else:\n",
    "        GT = \"0/0\"\n",
    "        \n",
    "    return GT\n",
    "    \n",
    "def mean_quality(reads, quals):\n",
    "    \"\"\"Calculate mean base quality of variant\"\"\"\n",
    "    # ?<!\\^ Negative lookbehind for ^ char and letters after it to calculate average base quality of found bases\n",
    "    alt_base_reg = re.compile(\"(?<!\\^)[ACTGNactgn]\")\n",
    "    altbasesIter = alt_base_reg.finditer(reads)\n",
    "    # Take indexes of bases quality which are found as mutations \n",
    "    alt_indexes = [alt_ind.start(0) for alt_ind in altbasesIter]\n",
    "    # Create list of qualities\n",
    "    qual_list = [qual for qual in quals]\n",
    "    # Phred Quality of those bases found in qual list, represented in ASCII with -33 offset\n",
    "    qualities = [(ord(qual_list[i]) - 33) for i in alt_indexes]\n",
    "    # Calculate mean quality of Phred qualities extracted\n",
    "    mean_qual = str(round(sum(qualities)/len(qualities),2))\n",
    "    \n",
    "    return mean_qual\n",
    "\n",
    "def split_line(line):\n",
    "    \"\"\"Split line by tab to important columns and extract important columns.\n",
    "    Indexes are taken as default with known pileup format.\"\"\"\n",
    "    parts = line.strip('\\n').split('\\t')\n",
    "    chrom = parts[0]\n",
    "    pos = parts[1]\n",
    "    ref_base = parts[2]\n",
    "    cov = int(parts[3])\n",
    "    reads = parts[-2].replace(\"$\", \"\")\n",
    "    qual_part = parts[-1]\n",
    "    \n",
    "    return chrom, pos, ref_base, cov, reads, qual_part\n",
    "    \n",
    "def quality_filter(mean_qual):\n",
    "    \"\"\"Filter variants by quality\"\"\"\n",
    "    mean_qual = float(mean_qual)\n",
    "    if mean_qual > qual_threshold:\n",
    "        filt = \"PASS\"\n",
    "    else:\n",
    "        filt = \"LowQual\"\n",
    "        \n",
    "    return filt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open mpileup for reading and VCF for writing\n",
    "# example task with pileup output file can be found here: \n",
    "# https://cgc.sbgenomics.com/u/ana_djukic/matf-2021/tasks/d8563fba-1576-4755-bdb5-e26251dcdab0/\n",
    "\n",
    "with open('/sbgenomics/project-files/ < FILE NAME >', 'r') as pileup, open('test-merged-normal.vcf', 'w') as vcf:\n",
    "    \n",
    "    # Add header of VCF by its specification\n",
    "    vcf.write(\"##fileformat=VCFv4.2\\n\")\n",
    "    \n",
    "    # Add header columns in VCF\n",
    "    vcf.write(\"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\\tSAMPLE\\n\")\n",
    "    \n",
    "    for line in pileup:\n",
    "        \n",
    "        # We split and return parts of each pileup line\n",
    "        chrom, pos, ref_base, cov, reads, qual_part = split_line(line)\n",
    "        \n",
    "        # If coverage larger than threshold then parse reads line\n",
    "        if cov > cov_threshold:\n",
    "            # Retrun alternative base and number of alternative alleles\n",
    "            alt, alt_allels = alternative_base(reads) \n",
    "            \n",
    "            # Count Variant allele frequency\n",
    "            vaf = calculate_vaf(alt_allels, cov)\n",
    "            \n",
    "            # If VAF is larger than threshold \n",
    "            # and length of bases in pileup is equal of length of qualities\n",
    "            if vaf > vaf_threshold and (len(reads) == len(qual_part)):\n",
    "                \n",
    "                # Return mean quality\n",
    "                mean_qual = mean_quality(reads,qual_part)\n",
    "                \n",
    "                # Return filter field\n",
    "                filt = quality_filter(mean_qual)\n",
    "                \n",
    "                # Return calculated GT \n",
    "                GT = genotype(vaf, het_threshold, hom_threshold)\n",
    "                \n",
    "                # Create sample column\n",
    "                sample_string = \":\".join([GT, str(vaf),str(cov), str(alt_allels)])\n",
    "                \n",
    "                # Create line which will be written in VCF\n",
    "                line_to_write = '\\t'.join([chrom, pos, '.', ref_base, alt, \n",
    "                                           mean_qual, filt, '.', \"GT:VAF:DP:AD\", \n",
    "                                           sample_string])\n",
    "                # Write line to VCF    \n",
    "                vcf.write(line_to_write + '\\n')\n",
    "        \n",
    "        # If coverage is lower than threshold skip this position            \n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
